src/plugins/diagram-viewer/diagram-content.ts

```diff
@@ -22,48 +24,89 @@ export const DiagramContentModel = ToolContentModel
+          // below. If we don't skip these steps we can get in an infinite 
+          // loop.
+        } else {
+          if (!containerSharedModel) {
+            // The document doesn't have a shared model yet
+            containerSharedModel = SharedVariables.create();
+            console.log("newly created shared model", getSnapshot(containerSharedModel));
+          } 
   
-        destroy(node);
-      }
-    });
+          // CHECKME: this might trigger an pre-mature update because the document's
+          // shared models will be updated first before the tiles. And the update to the
+          // document's shared models will trigger the document level autorun.
+          sharedModelManager.setTileSharedModel(self, "variables", containerSharedModel);  
 
-    return {removeVariable};
-  })
+        }
+        self.root.setVariablesAPI(containerSharedModel);
+      }, 
+      {name: "sharedModelSetup", fireImmediately: true}));
+    }
+  }))
   .actions(self => ({
-    afterAttach() {
+    updateAfterSharedModelChanges() {
+      // First cleanup any invalid references this can happen when a item is deleted
+      self.root.nodes.forEach(node => {
+        // If the sharedItem is not valid destroy the list item
+        if (!isValidReference(() => node.variable)) {
+          destroy(node);
+        }
+      });        
+    
       if (!self.sharedModel) {
-        self.sharedModel = SharedVariables.create();
+        // We should never get into this case, but this is here to just in case 
+        // somehow we do
+        return;
       }
-
-      // When the tree monitor code is in place we can just set the sharedModel 
-      // as the variablesAPI. And the updating of the node when a variable is deleted
-      // will be handled by an update function. 
-      // In the meantime we emulate this effect with an async removeVariable implementation
-      self.root.setVariablesAPI({
-        createVariable(): VariableType {
-          if (!self.sharedModel) {
-            throw new Error("Need a sharedModel to create variables");
-          }
-          return self.sharedModel.createVariable();
-        },
-        removeVariable(variable?: VariableType): void {
-          self.removeVariable(variable);
+  
+      Array.from(self.sharedModel.variables.values()).forEach(variable => {
+        // sync up shared data model items with the tile data of items
+        // look for this item in the itemList, if it is not there add it
+        const sharedItemId = variable.id;
+        
+        // the dereferencing of sharedItem should be safe here because we first cleaned up any
+        // items that referenced invalid shared items.
+        const nodes = Array.from(self.root.nodes.values());
+        const matchingItem = nodes.find(node => node.variable.id === sharedItemId);
+        if (!matchingItem) {
+          const newItem = DQNode.create({ variable: sharedItemId, x: 100, y: 100 });
```

Comment:
Eventually we'll want a better heuristic for initial placement. Could call a `getPositionForNewNode` function which just returns (100, 100) for now to make that clear.

Reply:
Done.